toolkits/slack/arcade_slack/utils.py

```diff
@@ -0,0 +1,201 @@
+        "real_name": user_info.get("real_name"),
+        "timezone": user_info.get("tz"),
+    }
+
+
+def is_user_a_bot(user: dict) -> bool:
+    """Check if a Slack user object represents a bot.
+
+    Bots are users with the "is_bot" flag set to true.
+    USLACKBOT is the user object for the Slack bot itself and is a special case.
+
+    See https://api.slack.com/types/user for the structure of the user object.
+    """
+    return user.get("is_bot") or user.get("id") == "USLACKBOT"
+
+
+def is_user_deleted(user: dict) -> bool:
+    """Check if a Slack user object represents a deleted user.
+
+    See https://api.slack.com/types/user for the structure of the user object.
+    """
+    return user.get("deleted", False)
+
+
+async def async_paginate(
+    func: Callable,
+    response_key: Optional[str] = None,
+    limit: Optional[int] = MAX_PAGINATION_LIMIT,
+    next_cursor: Optional[str] = None,
+    *args,
+    **kwargs,
+) -> tuple[list, Optional[str]]:
+    """Paginate a Slack AsyncWebClient's function results.
+
+    The purpose is to abstract the pagination work and make it easier for the LLM to retrieve the
+    amount of items requested by the user, regardless of limits imposed by the Slack API. We still
+    return the next cursor, if needed to paginate further.
+    """
+    results = []
+    while len(results) < limit:
+        slack_limit = min(limit - len(results), MAX_PAGINATION_LIMIT)
+        response = await func(*args, **{**kwargs, "limit": slack_limit, "cursor": next_cursor})
+
+        try:
+            results.extend(dict(response.data) if not response_key else response[response_key])
+        except KeyError:
+            raise ValueError(f"Response key {response_key} not found in Slack response")
+
+        next_cursor = response.get("response_metadata", {}).get("next_cursor")
+
+        if not next_cursor:
+            break
+
+    return results, next_cursor
+
+
+def enrich_message_metadata(message: dict) -> dict:
+    """Enrich message metadata."""
+    message = enrich_message_datetime(message)
+    return message
+
+
+def enrich_message_datetime(message: dict) -> dict:
+    """Enrich message metadata with formatted datetime.
+
+    It helps LLMs when they need to display the date/time in human-readable format. Slack
+    will only return a unix-formatted timestamp (it's not actually unix ts, but the unix ts
+    in the user's timezone - I know, odd, but it is what it is).
+    """
+    ts = message.get("ts")
+    if ts:
+        message["datetime_timestamp"] = datetime.fromtimestamp(float(ts)).strftime(
+            "%Y-%m-%d %H:%M:%S"
+        )
+    return message
+
+
+def convert_datetime_to_unix_timestamp(datetime_str: str) -> int:
+    """Convert a datetime string to a unix timestamp."""
+    try:
+        dt = datetime.strptime(datetime_str, "%Y-%m-%d %H:%M:%S")
+        return int(dt.timestamp())
+    except ValueError:
+        raise RetryableToolError(
```

Comment:
design question... should a ``RetryableToolError`` ever be raised outside a tool? maybe? thoughts?

Reply:
That got me thinking, too.

The reason I ended up doing it is because we'd nee to repeat this except block twice in a tool, making it unnecessarily longer and a bit harder to read. This makes the tool cleaner. 

I think it's a fair trade-off, and since those are tool util, not general-purpose, they're expected to be called within a tool execution runtime anyways...